torch/csrc/jit/testing/file_check.cpp

```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
+    std::vector<Check> operands;
+    const static std::vector<std::pair<CheckType, std::string>> check_pairs = {
+        {CHECK, ": "},
+        {CHECK_NEXT, "-NEXT: "},
+        {CHECK_SAME, "-SAME: "},
+        {CHECK_NOT, "-NOT: "},
+        {CHECK_DAG, "-DAG: "},
+        {CHECK_COUNT, "-COUNT-"}, // needs special parsing
+    };
+    const std::string prefix = "# CHECK";
+
+    size_t start = 0;
+    size_t num_checks = 0;
+    start = checks_file->find(prefix, start);
+
+    while (start != std::string::npos) {
+      bool found_match = false;
+      for (const auto& check_pair : check_pairs) {
+        const std::string& check_suffix = check_pair.second;
+        auto suffix_pos = checks_file->find(check_suffix, start);
+        if (suffix_pos != start + prefix.size()) {
+          continue;
+        }
+        size_t end_check_string = suffix_pos + check_suffix.size();
+        CheckType type = check_pair.first;
+        c10::optional<size_t> count = c10::nullopt;
+        auto end_line = checks_file->find("\n", end_check_string);
+        bool exactly = false;
+        if (type == CHECK_COUNT) {
+          const std::string exact = "EXACTLY-";
+          if (checks_file->find(exact, end_check_string) == end_check_string) {
+            exactly = true;
+            end_check_string += exact.size();
+          }
+          size_t end = assertFind(
+              SourceRange(checks_file, end_check_string, end_line), ":");
+          count = std::stoll(
+              checks_file->substr(end_check_string, end - end_check_string));
+          end_check_string = end + 2; // add ':' and the space
+        }
+        auto check = Check(
+            type,
+            checks_file->substr(end_check_string, end_line - end_check_string),
+            count);
+        addCheck(check);
+        if (exactly) {
+          addCheck(CHECK_NOT, check.search_str_);
+        }
+        start = end_line;
+        found_match = true;
+        break;
+      }
+      if (!found_match) {
+        std::ostringstream ss;
+        ss << "Could not parse check at:\n";
+        SourceRange(checks_file, start, start + prefix.size()).highlight(ss);
+        ss << "Check for bad input.";
+        has_run = true;
+        throw std::runtime_error(ss.str());
+      }
+
+      start = checks_file->find(prefix, start);
+      num_checks++;
+    }
+    size_t num_check_string = substringCount(checks_file, "CHECK");
+    if (num_checks != 0 && num_checks == num_check_string)
+      return;
+
+    std::ostringstream ss;
+    if (num_checks < num_check_string) {
+      ss << num_checks << " checks generated but found " << num_check_string
+         << " 'CHECK' "
+         << "in input string. Check for bad input:\n"
+         << *checks_file;
+    } else if (num_checks == 0) {
+      ss << "0 checks generated. Check for bad input:\n" << *checks_file;
+    }
+    has_run = true; // no extra print after exception thrown
```

Comment:
Do we actually need the `has_run` field? I see some value in verifying the user doesn't screw up, but at the same time I'd vote for maximum simplicity of FileCheck implementation, meaning removing non-critical checks.

Reply:
Sure i can remove it - there shouldn't be any tests cases checked in with bad inputs anyway (except for my test cases, which i can suppress output of)

Reply:
It doesn't have to be in this PR, and if you think it's worth having it, that's fine with me too.